File: v1/src/lenticularis/yamlconf.py
"""A conf file reader.  Conf files are in yaml, and they are read
and checked against json schema."""

# Copyright (c) 2022-2023 RIKEN R-CCS
# SPDX-License-Identifier: BSD-2-Clause

import sys
import json
import jsonschema
import yaml
from lenticularis.utility import rephrase_exception_message


def read_yaml_conf(file):
    """Reads a file and checks it against json schema."""
    assert file is not None
    try:
        with open(file, "r") as f:
            yamlconf = yaml.load(f, Loader=yaml.BaseLoader)
    except yaml.YAMLError as e:
        raise Exception(f"Reading a conf file failed: {file}:"
                        f" exception=({e})")
    except Exception as e:
        m = rephrase_exception_message(e)
        raise Exception(f"Reading a conf file failed: {file}:"
                        f" exception=({m})")
    if "subject" not in yamlconf:
        raise Exception(f"Bad conf file: {file}:"
                        f" missing subject")
    sub = yamlconf["subject"]
    if sub == "api":
        schema = _api_conf_schema()
        conf = _fix_type(yamlconf, schema)
        jsonschema.validate(instance=conf, schema=schema)
        return conf
    elif sub[:3] == "mux":
        schema = _mux_conf_schema()
        conf = _fix_type(yamlconf, schema)
        jsonschema.validate(instance=conf, schema=schema)
        return conf
    else:
        raise Exception(f"Bad conf file: {file}:"
                        f" bad subject=({sub})")
    pass


redis_json_schema = {
    "type": "object",
    "properties": {
        "host": {"type": "string"},
        "port": {"type": "number"},
        "password": {"type": "string"},
    },
    "required": [
        "host",
        "port",
        "password",
    ],
    "additionalProperties": False,
}

_gunicorn_json_schema = {
    "type": "object",
    "properties": {
        "port": {"type": "string"},
        "workers": {"type": "number"},
        "threads": {"type": "number"},
        "timeout": {"type": "number"},
        "access_logfile": {"type": "string"},
        "reload": {"type": "string"},
        "log_file": {"type": "string"},
        "log_level": {"type": "string"},
        "log_syslog_facility": {"type": "string"},
    },
    "required": [
        "port",
    ],
    "additionalProperties": False,
}

_syslog_json_schema = {
    "type": "object",
    "properties": {
        "facility": {"type": "string"},
        "priority": {"type": "string"},
    },
    "required": [
        "facility",
        "priority",
    ],
    "additionalProperties": False,
}

_minio_json_schema = {
    "type": "object",
    "properties": {
        "minio": {"type": "string"},
        "mc": {"type": "string"},
    },
    "required": [
        "minio",
        "mc",
    ],
    "additionalProperties": False,
}

def _mux_conf_schema():
    """mux_node_name and log_file are optional."""
    multiplexer = {
        "type": "object",
        "properties": {
            "front_host": {"type": "string"},
            "trusted_proxies": {"type": "array", "items": {"type": "string"}},
            "mux_ep_update_interval": {"type": "number"},
            "forwarding_timeout": {"type": "number"},
            "probe_access_timeout": {"type": "number"},
            "bad_response_delay": {"type": "number"},
            "mux_node_name": {"type": "string"},
        },
        "required": [
            "front_host",
            "trusted_proxies",
            "mux_ep_update_interval",
            "forwarding_timeout",
            "probe_access_timeout",
            "bad_response_delay",
        ],
        "additionalProperties": False,
    }
    minio_manager = {
        "type": "object",
        "properties": {
            "sudo": {"type": "string"},
            "port_min": {"type": "number"},
            "port_max": {"type": "number"},
            "minio_awake_duration": {"type": "number"},
            "minio_setup_at_restart": {"type": "boolean"},
            "heartbeat_interval": {"type": "number"},
            "heartbeat_miss_tolerance": {"type": "number"},
            "heartbeat_timeout": {"type": "number"},
            "minio_start_timeout": {"type": "number"},
            "minio_setup_timeout": {"type": "number"},
            "minio_stop_timeout": {"type": "number"},
            "minio_mc_timeout": {"type": "number"},
        },
        "required": [
            "sudo",
            "port_min",
            "port_max",
            "minio_awake_duration",
            "minio_setup_at_restart",
            "heartbeat_interval",
            "heartbeat_miss_tolerance",
            "heartbeat_timeout",
            "minio_start_timeout",
            "minio_setup_timeout",
            "minio_stop_timeout",
            "minio_mc_timeout",
        ],
        "additionalProperties": False,
    }
    _schema = {
        "type": "object",
        "properties": {
            "subject": {"type": "string"},
            "version": {"type": "string"},
            "aws_signature": {"type": "string"},
            "gunicorn": _gunicorn_json_schema,
            "multiplexer": multiplexer,
            "minio_manager": minio_manager,
            "minio": _minio_json_schema,
            "log_file": {"type": "string"},
            "log_syslog": _syslog_json_schema,
        },
        "required": [
            "subject",
            "version",
            "aws_signature",
            "gunicorn",
            "multiplexer",
            "minio_manager",
            "minio",
            "log_syslog",
        ],
        "additionalProperties": False,
    }
    return _schema


def _api_conf_schema():
    """log_file is optional."""
    controller = {
        "type": "object",
        "properties": {
            "front_host": {"type": "string"},
            "trusted_proxies": {"type": "array", "items": {"type": "string"}},
            "base_path": {"type": "string"},
            "claim_uid_map": {"type": "string",
                              "enum": ["id", "email-name", "map"]},
            "probe_access_timeout": {"type": "number"},
            "minio_mc_timeout": {"type": "number"},
            "max_pool_expiry": {"type": "number"},
            "csrf_secret_key": {"type": "string"},
        },
        "required": [
            "front_host",
            "trusted_proxies",
            "base_path",
            "claim_uid_map",
            "probe_access_timeout",
            "minio_mc_timeout",
            "max_pool_expiry",
            "csrf_secret_key",
        ],
        "additionalProperties": False,
    }
    _schema = {
        "type": "object",
        "properties": {
            "subject": {"type": "string"},
            "version": {"type": "string"},
            "aws_signature": {"type": "string"},
            "gunicorn": _gunicorn_json_schema,
            "controller": controller,
            "minio": _minio_json_schema,
            "log_file": {"type": "string"},
            "log_syslog": _syslog_json_schema,
        },
        "required": [
            "subject",
            "version",
            "aws_signature",
            "gunicorn",
            "controller",
            "minio",
            "log_syslog",
        ],
        "additionalProperties": False,
    }
    return _schema


def _fix_type(data, schema):
    """Rereads tokens in yaml to match for json schema.  It admits
    missing/additional properties, which will be checked by json
    validation.
    """
    if schema["type"] == "object":
        newdata = data.copy()
        for (prop, subschema) in schema["properties"].items():
            subdata = data.get(prop)
            # assert prop not in schema["required"] or subdata is not None
            if subdata is not None:
                newdata[prop] = _fix_type(subdata, subschema)
                pass
            pass
        return newdata
    elif schema["type"] == "array":
        subschema = schema["items"]
        assert isinstance(data, list)
        return [_fix_type(subdata, subschema) for subdata in data]
    elif schema["type"] == "string":
        assert isinstance(data, str)
        return data
    elif schema["type"] == "number":
        assert isinstance(data, str)
        try:
            return int(data)
        except ValueError:
            return float(data)
    elif schema["type"] == "boolean":
        assert isinstance(data, str)
        return bool(data)
    else:
        raise Exception("_fix_type: Other types are not implemented")
    pass

File: v2/pkg/lens3/yamlconf.go
/* A conf file reader. */

// Copyright 2022-2024 RIKEN R-CCS
// SPDX-License-Identifier: BSD-2-Clause

package lens3

// Conf files are in yaml, and they are read
// and checked against json schema.

// https://pkg.go.dev/gopkg.in/yaml.v3@v3.0.1
// == https://gopkg.in/yaml.v3
// https://pkg.go.dev/sigs.k8s.io/yaml

import (
	"encoding/json"
	"fmt"
	"gopkg.in/yaml.v3"
	"io/ioutil"
	//"reflect"
)

// Number representation in the configuration structure is integer.
type Number int

// Lens3_conf is a union of Mux_conf|Api_conf.
type Lens3_conf interface{ lens3_conf_union() }

func (Mux_conf) lens3_conf_union() {}
func (Api_conf) lens3_conf_union() {}

type Conf_header struct {
	Subject       string
	Version       string
	Aws_signature string
}

type Redis_conf struct {
	Host     string
	Port     Number
	Password string
}

// Mux_conf is a configuration of Mux.  mux_node_name and log_file are
// optional.
type Mux_conf struct {
	Conf_header
	//Subject       string
	//Version       string
	//Aws_signature string
	Gunicorn      Gunicorn_conf
	Multiplexer   Multiplexer_conf
	Minio_Manager Manager_conf
	Minio         Minio_conf
	Log_file      string
	Log_syslog    Syslog_conf
}

// Api_conf is a configuration of Api.  log_file is optional.
type Api_conf struct {
	Conf_header
	//Subject       string
	//Version       string
	//Aws_signature string
	Gunicorn   Gunicorn_conf
	Controller Controller_conf
	UI         UI_conf
	Minio      Minio_conf
	Log_file   string
	Log_syslog Syslog_conf
}

type Multiplexer_conf struct {
	Front_host             string
	Trusted_proxies        []string
	Mux_ep_update_interval Number
	Forwarding_timeout     Number
	Probe_access_timeout   Number
	Bad_response_delay     Number
	Busy_suspension_time   Number
	Mux_node_name          string
}

type Controller_conf struct {
	Front_host      string
	Trusted_proxies []string
	Base_path       string
	Claim_uid_map   string
	// {"id", "email-name", "map"}
	Probe_access_timeout Number
	Minio_mc_timeout     Number
	Max_pool_expiry      Number
	Csrf_secret_seed     string
}

type Manager_conf struct {
	Sudo                     string
	Port_min                 Number
	Port_max                 Number
	Minio_awake_duration     Number
	Minio_setup_at_start     bool
	Heartbeat_interval       Number
	Heartbeat_miss_tolerance Number
	Heartbeat_timeout        Number
	Minio_start_timeout      Number
	Minio_setup_timeout      Number
	Minio_stop_timeout       Number
	Minio_mc_timeout         Number
}

type Minio_conf struct {
	Minio string
	Mc    string
}

type UI_conf struct {
	S3_url        string
	Footer_banner string
}

type Syslog_conf struct {
	Facility string
	Priority string
}

type Gunicorn_conf struct {
	Port                Number
	Workers             Number
	Threads             Number
	Timeout             Number
	Access_logfile      string
	Reload              string //bool
	Log_file            string
	Log_level           string
	Log_syslog_facility string
}

// Read_yaml_conf reads a configuration file and checks a structure is
// properly filled.
func Read_yaml_conf(file_ string) interface{} {
	var b1, err1 = ioutil.ReadFile("conf.yaml")
	if err1 != nil {
		panic(err1)
	}
	var yaml2 = make(map[string]interface{})
	var err2 = yaml.Unmarshal(b1, &yaml2)
	if err2 != nil {
		panic(err2)
	}
	var json3, err4 = json.Marshal(yaml2)
	if err4 != nil {
		panic(err4)
	}

	//fmt.Println(yaml2)
	//fmt.Println(string(json3))

	/*
		var c2 = yaml2["controller"].(map[string]interface{})
		fmt.Println(c2)
		var t2 = c2["minio_mc_timeout"]
		fmt.Println(reflect.TypeOf(t2))
		fmt.Println(t2)
	*/
	/*
		var b3, err3 = json.Marshal(&yaml2)
		if err3 != nil {
			panic(err3)
		}
		fmt.Print(b3)
	*/
	/*
		fmt.Println(yaml2)
	*/

	var sub = yaml2["subject"].(string)[:3]
	switch sub {
	case "mux":
		var muxconf Mux_conf
		var err5 = json.Unmarshal(json3, &muxconf)
		if err5 != nil {
			panic(fmt.Sprint("Bad .yaml conf file:", err5))
		}
		fmt.Println("MUX CONF is", muxconf)
		check_mux_conf(muxconf)
		return muxconf
	case "api":
		var apiconf Api_conf
		var err6 = json.Unmarshal(json3, &apiconf)
		if err6 != nil {
			panic(fmt.Sprint("Bad .yaml conf file:", err6))
		}
		fmt.Println("API CONF is", apiconf)
		check_api_conf(apiconf)
		return apiconf
	default:
		panic(fmt.Sprint("Bad .yaml conf file: Bad subject field."))
	}
}

func check_mux_conf(conf Mux_conf) {
	check_gunicorn_entry(conf.Gunicorn)
	check_multiplexer_entry(conf.Multiplexer)
	check_manager_entry(conf.Minio_Manager)
	check_minio_entry(conf.Minio)
	check_syslog_entry(conf.Log_syslog)
}

func check_api_conf(conf Api_conf) {
	check_gunicorn_entry(conf.Gunicorn)
	check_controller_entry(conf.Controller)
	check_minio_entry(conf.Minio)
	check_ui_entry(conf.UI)
	check_syslog_entry(conf.Log_syslog)
}

func assert_slot(c bool) {
	if !c {
		panic(fmt.Errorf("Bad .yaml conf file."))
	}
}

func check_gunicorn_entry(e Gunicorn_conf) {
	assert_slot(e.Port > 0)
	assert_slot(e.Workers > 0)
	//assert_slot(e.Threads >= 0)
	assert_slot(e.Timeout > 0)
	assert_slot(len(e.Access_logfile) > 0)
	// assert_slot(e.Reload is bool and always OK.)
	assert_slot(len(e.Log_file) > 0)
	assert_slot(len(e.Log_level) > 0)
	assert_slot(len(e.Log_syslog_facility) >= 0)
}

func check_controller_entry(e Controller_conf) {
	if len(e.Front_host) > 0 &&
		//e.Trusted_proxies : []string
		len(e.Base_path) > 0 &&
		len(e.Claim_uid_map) > 0 &&
		// {"id", "email-name", "map"}
		e.Probe_access_timeout > 0 &&
		e.Minio_mc_timeout > 0 &&
		e.Max_pool_expiry > 0 &&
		len(e.Csrf_secret_seed) > 0 {
	} else {
		panic(fmt.Errorf("Bad .yaml conf file."))
	}
}

func check_multiplexer_entry(e Multiplexer_conf) {
	if len(e.Front_host) > 0 &&
		//e.Trusted_proxies : []string
		e.Mux_ep_update_interval > 0 &&
		e.Forwarding_timeout > 0 &&
		e.Probe_access_timeout > 0 &&
		e.Bad_response_delay > 0 &&
		e.Busy_suspension_time > 0 &&
		len(e.Mux_node_name) >= 0 {
	} else {
		panic(fmt.Errorf("Bad .yaml conf file."))
	}
}

func check_manager_entry(e Manager_conf) {
	assert_slot(len(e.Sudo) > 0)
	assert_slot(e.Port_min > 0)
	assert_slot(e.Port_max > 0)
	assert_slot(e.Minio_awake_duration > 0)
	//assert_slot(Minio_setup_at_start : bool
	assert_slot(e.Heartbeat_interval > 0)
	assert_slot(e.Heartbeat_miss_tolerance > 0)
	assert_slot(e.Heartbeat_timeout > 0)
	assert_slot(e.Minio_start_timeout > 0)
	assert_slot(e.Minio_setup_timeout > 0)
	assert_slot(e.Minio_stop_timeout > 0)
	assert_slot(e.Minio_mc_timeout > 0)
}

func check_redis_entry(e Redis_conf) {
	if len(e.Host) > 0 &&
		e.Port > 0 &&
		len(e.Password) > 0 {
	} else {
		panic(fmt.Errorf("Bad .yaml conf file."))
	}
}

func check_minio_entry(e Minio_conf) {
	if len(e.Minio) > 0 &&
		len(e.Mc) > 0 {
	} else {
		panic(fmt.Errorf("Bad .yaml conf file."))
	}
}

func check_ui_entry(e UI_conf) {
	if len(e.S3_url) > 0 &&
		len(e.Footer_banner) > 0 {
	} else {
		panic(fmt.Errorf("Bad .yaml conf file."))
	}
}

func check_syslog_entry(e Syslog_conf) {
	if len(e.Facility) > 0 &&
		len(e.Priority) > 0 {
	} else {
		panic(fmt.Errorf("Bad .yaml conf file."))
	}
}
